File: typescript/src/infrastructure/split-on-go.ts
const GO_LINE = /^\s*GO\s*$/i;

/** Split migration SQL on whole-line GO batch separators. */
export function splitOnGo(text: string): string[] {
  const batches: string[] = [];
  let buf = '';
  for (const line of text.replace(/\r\n/g, '\n').split('\n')) {
    if (GO_LINE.test(line)) {
      const s = buf.trim();
      if (s.length > 0) batches.push(s);
      buf = '';
      continue;
    }
    buf = buf.length > 0 ? `${buf}\n${line}` : line;
  }
  const tail = buf.trim();
  if (tail.length > 0) batches.push(tail);
  return batches;
}

File: rust/src/infrastructure/split_on_go.rs
// Whole-line GO separates batches (sqlcmd-style).
pub fn split_on_go(sql: &str) -> Vec<String> {
    let mut out = Vec::new();
    let mut buf = String::new();
    for line in sql.replace("\r\n", "\n").split('\n') {
        if line.trim().eq_ignore_ascii_case("GO") {
            let s = buf.trim();
            if !s.is_empty() {
                out.push(s.to_string());
            }
            buf.clear();
            continue;
        }
        if !buf.is_empty() {
            buf.push('\n');
        }
        buf.push_str(line);
    }
    let tail = buf.trim();
    if !tail.is_empty() {
        out.push(tail.to_string());
    }
    out
}

#[cfg(test)]
mod tests {
    use super::split_on_go;

    #[test]
    fn splits_on_whole_line_go() {
        assert_eq!(
            split_on_go("CREATE TABLE a (id INT);\nGO\nCREATE TABLE b (id INT)"),
            vec![
                "CREATE TABLE a (id INT);".to_string(),
                "CREATE TABLE b (id INT)".to_string(),
            ]
        );
    }

    #[test]
    fn keeps_semicolons_inside_one_batch() {
        assert_eq!(
            split_on_go("BEGIN\n  SELECT 1;\nEND;"),
            vec!["BEGIN\n  SELECT 1;\nEND;".to_string()]
        );
    }
}
